File: core/src/test/java/org/elkd/core/consensus/DefaultRaftDelegateFactoryTest.java
package org.elkd.core.consensus;

import org.elkd.core.config.ConfigProvider;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.doReturn;

public class DefaultRaftDelegateFactoryTest {

  @Mock Raft mRaft;

  private DefaultStateFactory mUnitUnderTest;

  @Before
  public void setup() throws Exception {
    MockitoAnnotations.initMocks(this);

    mUnitUnderTest = new DefaultStateFactory();

    doReturn(ConfigProvider.compileConfig(new String[] { }))
        .when(mRaft)
        .getConfig();
  }

  @Test
  public void should_create_singleton_instances() {
    // Given
    final State follower = mUnitUnderTest.getState(mRaft, RaftFollowerState.class);
    final State candidate = mUnitUnderTest.getState(mRaft, RaftCandidateState.class);
    final State leader = mUnitUnderTest.getState(mRaft, RaftLeaderState.class);

    // When
    final State otherFollower = mUnitUnderTest.getState(mRaft, RaftFollowerState.class);
    final State otherCandidate = mUnitUnderTest.getState(mRaft, RaftCandidateState.class);
    final State otherLeader = mUnitUnderTest.getState(mRaft, RaftLeaderState.class);

    // Then
    assertSame(follower, otherFollower);
    assertSame(candidate, otherCandidate);
    assertSame(leader, otherLeader);
  }

  @Test
  public void should_return_initial_state() {
    // Given
    final Class<? extends State> initialStateClass = DefaultStateFactory.INITIAL_STATE;

    // When
    final State initialDelegate = mUnitUnderTest.getInitialDelegate(mRaft);

    // Then
    assertTrue(initialStateClass.isInstance(initialDelegate));
  }
}

File: core/src/test/java/org/elkd/core/consensus/DefaultRaftDelegateFactoryTest.kt
package org.elkd.core.consensus

import kotlin.test.assertTrue
import org.elkd.core.config.ConfigProvider
import org.elkd.core.consensus.states.RaftStateFactory
import org.elkd.core.consensus.states.State
import org.elkd.core.consensus.states.candidate.RaftCandidateState
import org.elkd.core.consensus.states.follower.RaftFollowerState
import org.elkd.core.consensus.states.leader.RaftLeaderState
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.MockitoAnnotations

class DefaultRaftDelegateFactoryTest {

  @Mock internal lateinit var raft: Raft

  private var unitUnderTest: RaftStateFactory? = null

  @Before
  @Throws(Exception::class)
  fun setup() {
    MockitoAnnotations.initMocks(this)

    unitUnderTest = RaftStateFactory(raft)

    doReturn(ConfigProvider.compileConfig(arrayOf()))
        .`when`(raft)
        .config
  }

  @Test
  fun should_return_correct_states() {
    // Given / When
    val followerState = unitUnderTest?.getState(State.FOLLOWER)
    val candidateState = unitUnderTest?.getState(State.CANDIDATE)
    val leaderState = unitUnderTest?.getState(State.LEADER)

    // Then
    assertTrue { followerState is RaftFollowerState }
    assertTrue { candidateState is RaftCandidateState }
    assertTrue { leaderState is RaftLeaderState }
  }
}
